rust/worker/src/blockstore/arrow_blockfile/block/delta.rs

```diff
@@ -0,0 +1,382 @@
+use super::{Block, BlockBuilderOptions, BlockData, BlockDataBuilder};
+use crate::blockstore::{
+    arrow_blockfile::{blockfile::MAX_BLOCK_SIZE, provider::ArrowBlockProvider},
+    types::{BlockfileKey, KeyType, Value, ValueType},
+};
+use arrow::util::bit_util;
+use parking_lot::RwLock;
+use std::{collections::BTreeMap, sync::Arc};
+
+/// A block delta tracks a source block and represents the new state of a block. Blocks are
+/// immutable, so when a write is made to a block, a new block is created with the new state.
+/// A block delta is a temporary representation of the new state of a block. A block delta
+/// can be converted to a block data, which is then used to create a new block. A block data
+/// can be converted into a block delta for new writes.
+/// # Methods
+/// - can_add: checks if a key value pair can be added to the block delta and still be within the
+///  max block size.
+/// - add: adds a key value pair to the block delta.
+/// - delete: deletes a key from the block delta.
+/// - get_min_key: gets the minimum key in the block delta.
+/// - get_size: gets the size of the block delta.
+/// - split: splits the block delta into two block deltas.
+#[derive(Clone)]
+pub struct BlockDelta {
+    pub source_block: Arc<Block>,
+    inner: Arc<RwLock<BlockDeltaInner>>,
+}
+
+impl BlockDelta {
+    /// Checks if a key value pair can be added to the block delta and still be within the
+    /// max block size.
+    pub fn can_add(&self, key: &BlockfileKey, value: &Value) -> bool {
+        let inner = self.inner.read();
+        inner.can_add(key, value)
+    }
+
+    /// Adds a key value pair to the block delta.
+    pub fn add(&self, key: BlockfileKey, value: Value) {
+        let mut inner = self.inner.write();
+        inner.add(key, value);
+    }
+
+    /// Deletes a key from the block delta.
+    pub fn delete(&self, key: BlockfileKey) {
+        let mut inner = self.inner.write();
+        inner.delete(key);
+    }
+
+    /// Gets the minimum key in the block delta.
+    pub fn get_min_key(&self) -> Option<BlockfileKey> {
+        let inner = self.inner.read();
+        let first_key = inner.new_data.keys().next();
+        first_key.cloned()
+    }
+
+    /// Gets the size of the block delta as it would be in a block. This includes
+    /// the size of the prefix, key, and value data and the size of the offsets
+    /// where applicable. The size is rounded up to the nearest 64 bytes as per
+    /// the arrow specification. When a block delta is converted into a block data
+    /// the same sizing is used to allocate the memory for the block data.
+    pub fn get_size(&self) -> usize {
+        let inner = self.inner.read();
+        inner.get_size(
+            self.source_block.get_key_type(),
+            self.source_block.get_value_type(),
+        )
+    }
+
+    /// Splits the block delta into two block deltas. The split point is the last key
+    /// that pushes the block over the half size.
```

Comment:
This returns a key and the new, split BlockDelta containing values strictly greater than values in this block, right? Could we add that to this comment